gcovr/configuration.py

```diff
@@ -0,0 +1,475 @@
+# -*- coding:utf-8 -*-
+
+# This file is part of gcovr <http://gcovr.com/>.
+#
+# This module manages processing and validation of the configuration options
+# passed into gcovr.
+#
+#
+# Copyright 2013-2018 the gcovr authors
+# Copyright 2013 Sandia Corporation
+# This software is distributed under the BSD license.
+from argparse import ArgumentTypeError, SUPPRESS
+from locale import getpreferredencoding
+from multiprocessing import cpu_count
+import os
+
+
+def check_percentage(value):
+    r""""
+    Check that the percentage is within a reasonable range and if so return it.
+    """
+    try:
+        x = float(value)
+        if not (0.0 <= x <= 100.0):
+            raise ValueError()
+    except ValueError:
+        raise ArgumentTypeError(
+            "{value} not in range [0.0, 100.0]".format(value=value))
+    return x
+
+
+def check_non_empty(value):
+    if not value:
+        raise ArgumentTypeError("value should not be empty")
+    return value
+
+
+class GcovrConfigOption:
+    r"""
+    Represents a single setting for a gcovr runtime parameter.
+
+    Gcovr can be extensively configured through a series of options,
+    representing these options as a simple class object allows them to be
+    portabilty re-used in multiple configuration schemes. This is implemented
+    in a way similar to how options are defined in argparse. The converter
+    keyword argument is expected to return a valid conversion of a string
+    value or throw an error.
+
+    This class is initializer in a very similar fashion as adding arguments
+    to an argparse Parser instance with a few minor changes.
+        1. The first positional argument is the "name" (i.e. "dest" in
+           argparse speak) and is required.
+        2. A "group" keyword argument has been added which corresponds to a
+           key in the GCOVR_CONFIG_OPTION_GROUPS dictionary.
+    """
+
+    def __init__(self, name, *flags, **kwargs):
```

Comment:
Do we have to use `**kwargs` here, or can we enumerate all supported option-fields? Due to Python-2 restrictions we can't really combine `*flags` varargs with explicit keyword-only arguments, so we'd have to pass in an explicit list/tuple of flags. Something like:

``` python
def __init__(self, name, flags=(), group=None, help=None, action=None):

...
GcovrConfigOption('foo', ['--foo'], help="frobnicate something", action='store_true')
```

… and so on, where everything after `flags` is intended to be used as a keyword-argument. This would allow us to get rid of some `hasattr()/getattr()` calls later, and clarifies the intended API.